libbpfgo.go

```diff
@@ -489,140 +488,314 @@ type BPFMapCreateOpts struct {
-	opts.btf_key_type_id = C.uint(createOpts.BtfKeyTypeID)
-	opts.btf_value_type_id = C.uint(createOpts.BtfValueTypeID)
-	opts.btf_vmlinux_value_type_id = C.uint(createOpts.BtfVmlinuxValueTypeID)
-	opts.inner_map_fd = C.uint(createOpts.InnerMapFD)
-	opts.map_flags = C.uint(createOpts.MapFlags)
-	opts.map_extra = C.ulonglong(createOpts.MapExtra)
-	opts.numa_node = C.uint(createOpts.NumaNode)
-	opts.map_ifindex = C.uint(createOpts.MapIfIndex)
-
-	return &opts
-}
-
 // CreateMap creates a BPF map from userspace. This can be used for populating
 // BPF array of maps or hash of maps. However, this function uses a low-level
 // libbpf API; maps created in this way do not conform to libbpf map formats,
-// and therefore do not have access to libbpf high level bpf_map__* APIS
-// which causes different behavior from maps created in the kernel side code
+// and therefore do not have access to libbpf high level `bpf_map__*` APIS
+// which causes different behavior from maps created in the kernel side code.
 //
-// See usage of `bpf_map_create()` in kernel selftests for more info
+// See usage of `bpf_map_create()` in kernel selftests for more info.
 func CreateMap(mapType MapType, mapName string, keySize, valueSize, maxEntries int, opts *BPFMapCreateOpts) (*BPFMap, error) {
-	cs := C.CString(mapName)
-	fdOrError := C.bpf_map_create(uint32(mapType), cs, C.uint(keySize), C.uint(valueSize), C.uint(maxEntries), bpfMapCreateOptsToC(opts))
-	C.free(unsafe.Pointer(cs))
-	if fdOrError < 0 {
-		return nil, fmt.Errorf("could not create map: %w", syscall.Errno(-fdOrError))
+	cMapName := C.CString(mapName)
+	defer C.free(unsafe.Pointer(cMapName))
+
+	var cOpts *C.struct_bpf_map_create_opts
+	var errno error
+	if opts != nil {
+		cOpts, errno = C.bpf_map_create_opts_new(
+			C.uint(opts.BtfFD),
+			C.uint(opts.BtfKeyTypeID),
+			C.uint(opts.BtfValueTypeID),
+			C.uint(opts.BtfVmlinuxValueTypeID),
+			C.uint(opts.InnerMapFD),
+			C.uint(opts.MapFlags),
+			C.ulonglong(opts.MapExtra),
+			C.uint(opts.NumaNode),
+			C.uint(opts.MapIfIndex),
+		)
+		defer C.bpf_map_create_opts_free(cOpts)
+
+		if cOpts == nil {
+			return nil, fmt.Errorf("failed to create bpf_map_create_opts: %w", errno)
+		}
+	}
+
+	var fd C.int
+	fd, errno = C.bpf_map_create(uint32(mapType), cMapName, C.uint(keySize), C.uint(valueSize), C.uint(maxEntries), cOpts)
+	if fd < 0 {
+		return nil, fmt.Errorf("could not create map %s: %w", mapName, errno)
+	}
+
+	info, err := getMapInfoByFD(fd)
+	if err != nil {
+		return nil, err
 	}
 
 	return &BPFMap{
-		name:   mapName,
-		fd:     fdOrError,
+		name:   info.Name, // get the name assigned by the kernel
+		fd:     fd,
+		bpfMap: nil,
 		module: nil,
+	}, nil
+}
+
+// CreateMapBasedOn creates a BPF map based on an existing map. which is
+// convenient for avoiding the need to specify all the map attributes when
+// creating a new map.
+//
+// The `innerMapFD` is the file descriptor of the inner map for the map being
+// created. If it does not exist, use -1.
```

Comment:
Not sure this function should exist unless there is a very good use case for it (not the current one). Also, "inner" terminology does not make much sense if we think of this function `per se`.

Reply:
Just sugar, but we can remove it and use CreateMap directly. :+1: